### PR TITLE
DS-4056: Upgrade to Commons Configuration v2 (and Commons Lang v3)

### PR DESCRIPTION
**Pull request by tdonohue**
https://jira.duraspace.org/browse/DS-4056

This PR does the following:
1. Upgrades us to Apache Commons Configuration v2
2. Replace our custom Spring XML config integration (DSpaceConfigurationFactoryBean) with the new, out-of-the-box ConfigurationPropertiesFactoryBean. This is the same functionality that allows us to reference configurations (e.g. `${config}`) from Spring XML config files. See: https://github.com/DSpace/DSpace/commit/82ab2d858dbe6b9714692c1ccb4ffe53298075c3
3. Configures a new `DSpaceConfigurationPlaceholderConfigurer`, which allows us to also reference configurations via `@Value` or `@PropertySource` Spring annotations, e.g. `@Value("${config}")`.  This is a new feature in Commons Configuration v2. See: https://github.com/DSpace/DSpace/commit/2b294ffeb7e02fbb622c8f76e78d4f5f25c86821
4. Create/Enhance our Configuration Unit Tests to ensure identical functionality to current configuration. This includes creation of new tests to prove Spring integration works, and a test to prove that reloadable configs work.
5. Finally, also upgrades us from `commons-lang` to `commons-lang3`.  This is a self-contained commit (https://github.com/DSpace/DSpace/pull/2244/commits/d5602ea4ecaccd4acc66e9b89563f75b1fd4628e) and I can move it to a separate PR if needed.  I've included it here though as the commons-configuration2 upgrade resulted in dependency convergence issues around commons-lang (which could be easily cleaned up via an upgrade).  (NOTE: This upgrade was almost entirely a simple find & replace of imports)

This is ready for review and has new unit tests to prove upgrade works. All existing unit tests also succeed.

NOTE: While this touches a lot of files, most updates are very simple. The more significant changes are in the `DSpaceConfigurationService` class (and the new tests to prove the changes work).

### CODE REVIEW

**Comment by mwoodiupui** [comment]
This looks huge, but the bulk of the changes are changing "lang" to "lang3".

**Comment by tdonohue** [comment]
@mwoodiupui : correct, the vast majority of the changes here are changes of package name or POM changes (which can be skimmed rather easily). Actual code changes are limited to the `DSpaceConfigurationService` class, and unit tests to prove the upgrade worked. There may be a few other minor refactors sprinkled around to support the upgrade, but all should be straightforward.

Again, apologies for doing this all in one PR, it does look large, but I think reviewers will find it rather easy to review (as most changes can be quickly skimmed).

File: dspace-services/src/main/resources/spring/spring-dspace-core-services.xml
Comment (mwoodiupui) [inline dspace-services/src/main/resources/spring/spring-dspace-core-services.xml]: Do we need this *and* PropertySourcesPlaceholderConfigurer (which is meant to replace it)?

**Review by mwoodiupui — APPROVED** [review]
+1 Overall this looks right.  I have a couple of questions which should not block merging.

File: dspace-services/src/main/resources/spring/spring-dspace-core-services.xml
Comment (tdonohue) [reply dspace-services/src/main/resources/spring/spring-dspace-core-services.xml]: Good question. I actually overlooked the [PropertyPlaceholderConfigurer](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/beans/factory/config/PropertyPlaceholderConfigurer.html) now has a reference to "use [PropertySourcesPlaceholderConfigurer](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/context/support/PropertySourcesPlaceholderConfigurer.html)" as a replacement.

You are correct that I've configured *both* here.  I'll try out removing the former, and see if all our tests still pass (which should tell us if we actually need both here or not).  As you likely saw, I have separate unit tests that already test for the features provided by these `Configurer` classes are "working" properly with our ConfigurationService.

File: pom.xml
Comment (tdonohue) [reply pom.xml]: I'll give the build a try without this `exclusion` and see what happens.  I feel like it still may have been pulling in a *slightly different* commons-lang3 version (e.g. the subminor number was different), but I'll check to be sure.

Comment (tdonohue) [reply pom.xml]: Turns out this exclusion is seemingly not needed afterall (good eye!). Removed it.

File: dspace-services/src/main/resources/spring/spring-dspace-core-services.xml
Comment (tdonohue) [reply dspace-services/src/main/resources/spring/spring-dspace-core-services.xml]: Good eye here, We can remove the PropertyPlaceholderConfigurer entirely, and *just use* the PropertySourcesPlaceholderConfigurer.  I've done that in the latest commit & all tests pass.

File: dspace-api/src/main/java/org/dspace/statistics/Dataset.java
Comment (benbosman) [inline dspace-api/src/main/java/org/dspace/statistics/Dataset.java]: Why has this type been modified from float to String?

**Review by benbosman — APPROVED** [review]
I don't see any issues with this PR, but I was only wondering why the Dataset class's syntax was altered.

File: dspace-api/src/main/java/org/dspace/statistics/Dataset.java
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/statistics/Dataset.java]: @benbosman : because of the line immediately following that one... notice how the previous code takes in a `float` double array, clones it and directly casts it into a `String` double array.  

My compiler complained this cast was not allowed in Commons Lang v3.  I initially worked to find a replacement, but couldn't find an easy solution. After realizing that this `Dataset` constructor is unused, I decided to change the type instead.

If we ever find a real need for a `float` based constructor, we can revisit this. But currently we don't use or need it.

**Comment by tdonohue** [comment]
I just rebased this to resolve the minor merge conflict against `master`.  As it's now at +2, once Travis rechecks it, I'll go ahead and merge.

**Review by terrywbrady — APPROVED** [review]
Code looks good.  It is good to see the possibility of defining reloadable properties.  I assume that is a new feature.

**Pull request merged**
